core/types/subtyping.cc

```diff
@@ -470,6 +454,32 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
     return lubGround(gs, t1, t2);
 }
 
+TypePtr lubProxyNonProxy(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2) {
+    bool allowProxyInLub = isa_type<TupleType>(t1) || isa_type<ShapeType>(t1);
+    TypePtr und = t1.underlying(gs);
+    if (Types::isSubType(gs, und, t2)) {
+        return t2;
+    } else if (allowProxyInLub) {
+        auto *orType = cast_type<OrType>(t2);
+        if (orType != nullptr) {
```

Comment:
What we really want to do in this block is produce an OrType tree replacing a node with t1, if t1 is a supertype of that node. This is where a refactor like #4496 would be very handy.

Reply:
Using `lubDistributeOr` to do the thing.